libbeat/api/server.go

```diff
@@ -19,86 +19,84 @@ package api
-			return
-		}
-
-		w.Header().Set("Content-Type", "application/json; charset=utf-8")
-
-		data := monitoring.CollectStructSnapshot(monitoring.GetNamespace("info").GetRegistry(), monitoring.Full, false)
-
-		print(w, data, r.URL)
+// New creates a new API Server.
+func New(log *logp.Logger, mux *http.ServeMux, config *common.Config) (*Server, error) {
+	if log == nil {
+		log = logp.NewLogger("")
 	}
-}
 
-// stateHandler reports state metrics
-func stateHandler(w http.ResponseWriter, r *http.Request) {
-	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	cfgwarn.Experimental("Metrics endpoint is enabled.")
+	cfg := DefaultConfig
+	err := config.Unpack(&cfg)
+	if err != nil {
+		return nil, err
+	}
 
-	data := monitoring.CollectStructSnapshot(monitoring.GetNamespace("state").GetRegistry(), monitoring.Full, false)
+	l, err := makeListener(cfg.Host, cfg.Port)
+	if err != nil {
+		return nil, err
+	}
 
-	print(w, data, r.URL)
+	return &Server{mux: mux, l: l, config: cfg, log: log.Named("api")}, nil
 }
 
-// statsHandler report expvar and all libbeat/monitoring metrics
-func statsHandler(w http.ResponseWriter, r *http.Request) {
-	w.Header().Set("Content-Type", "application/json; charset=utf-8")
-
-	data := monitoring.CollectStructSnapshot(monitoring.GetNamespace("stats").GetRegistry(), monitoring.Full, false)
+// Start starts the HTTP server and accepting new connection.
+func (s *Server) Start() {
+	s.log.Info("Starting stats endpoint")
+	go func(l net.Listener) {
+		s.log.Infof("Metrics endpoint listening on: %s (configured: %s)", l.Addr().String(), s.config.Host)
+		http.Serve(l, s.mux)
+		s.log.Infof("Finished starting stats endpoint: %s", l.Addr().String())
+	}(s.l)
+}
 
-	print(w, data, r.URL)
+// Stop stops the API server and free any resource associated with the process like unix sockets.
+func (s *Server) Stop() error {
+	return s.l.Close()
 }
 
-func datasetHandler(w http.ResponseWriter, r *http.Request) {
-	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+func parse(host string, port int) (string, string, error) {
+	url, err := url.Parse(host)
+	if err != nil {
+		return "", "", err
+	}
 
-	data := monitoring.CollectStructSnapshot(monitoring.GetNamespace("dataset").GetRegistry(), monitoring.Full, false)
+	// When you don't explicitely define the Scheme we fallback on tcp + host.
+	if len(url.Host) == 0 && len(url.Scheme) == 0 {
+		addr := host + ":" + strconv.Itoa(port)
+		return "tcp", addr, nil
+	}
 
-	print(w, data, r.URL)
+	switch url.Scheme {
+	case "http":
+		return "tcp", url.Host, nil
+	case "unix":
+		return url.Scheme, url.Path, nil
+	default:
+		return "", "", fmt.Errorf("unknown scheme %s for host string %s", url.Scheme, host)
+	}
 }
 
-func print(w http.ResponseWriter, data common.MapStr, u *url.URL) {
-	query := u.Query()
-	if _, ok := query["pretty"]; ok {
-		fmt.Fprintf(w, data.StringToPrint())
-	} else {
-		fmt.Fprintf(w, data.String())
-	}
+func isWindows() bool {
```

Comment:
Should this be moved into `server_test.go`? Seems like it's the only place that really needs it?